lusidtools/cocoon/cocoon_printer.py

```diff
@@ -0,0 +1,159 @@
+import pandas as pd
+import numpy as np
+
+
+def check_errors_and_success(target_dict: dict, target_name: str, required_keys: list):
+    """
+    This function checks that a named dictionary contains a list of required keys
+    :param dict target_dict: Dictionary to check for keys in
+    :param str target_name: Variable name of dictionary
+    :param list required_keys: List of required keys
+    :return:
+    """
+    for key in required_keys:
+        if key not in target_dict.keys():
+            raise ValueError(f"'{key}' missing from {target_name}. {target_name} must include 'errors', and 'success' "
+                             f"as keys")
+
+
+def get_errors_from_response(list_of_API_exceptions):
+    """
+    This function gets the status code and reason from API exception
+    :param list_of_API_exceptions:
+    :return:
+    """
+    # get API exception
+
+    status = [item.status for item in list_of_API_exceptions]
+    reason = [item.reason for item in list_of_API_exceptions]
+    items_error = np.array([reason, status]).T.tolist()
+    errors = pd.DataFrame(items_error, columns=["error_items", "status"])
+
+    return errors
+
+
+def check_for_failed_items(list_of_responses: list):
+    """
+    Return True if list contains objects with failed attributes 
+    :param list list_of_responses: list of either UpsertInstrumentsResponse or UpsertQuotesResponse
+    :return bool: failed items exist in list
+    """
+    return len(list_of_responses) > 0 and len(list_of_responses[0].failed)
+
+
+def format_instruments_response(response: dict) -> (
+        pd.DataFrame, pd.DataFrame, pd.DataFrame):
+    """
+    This function unpacks a response from instrument requests and returns successful, failed and errored statuses for
+    request constituents.
+    :param dict response: response from Lusid-python-tools
+    :return: pd.DataFrame: Successful calls from request
+    :return: pd.DataFrame: Error responses from request that fail (APIExceptions: 400 errors)
+    :return: pd.DataFrame: Failed responses that LUSID rejected
+    """
+    file_type = "instruments"
+    check_errors_and_success(response[file_type], f"Response from {file_type} request", ["errors", "success"])
+
+    items_success, items_failed, items_error = [], [], []
+
+    # get success and failures
+    items_success = [batch.values[key].name for batch in response[file_type]["success"] for key in batch.values.keys()]
+
+    if check_for_failed_items(response[file_type]["success"]):
+        items_failed = [batch.failed[key].name for batch in response[file_type]["success"] for key in
+                        batch.failed.keys()]
```

Comment:
You should be able to do this even if failed is empty {}. That way you don't need to define items_failed = [] above as it will be defined in-line here via list comprehension.